src/NHibernate/Impl/FutureBatch.cs

```diff
@@ -80,5 +116,88 @@ private IEnumerable<TResult> GetCurrentResult<TResult>(int currentIndex)
 		protected abstract void ClearCurrentFutureBatch();
 		protected abstract bool IsQueryCacheable(TQueryApproach query);
 		protected abstract string CacheRegion(TQueryApproach query);
+
+		protected virtual void AddResultTransformer(
+			TMultiApproach multiApproach,
+			IResultTransformer futureResulsTransformer)
+		{
+			// Only Linq set ExecuteOnEval, so only FutureQueryBatch needs to support it, not FutureCriteriaBatch.
+			throw new NotSupportedException();
+		}
+
+		[Serializable]
+		private class BatchedQueryPostExecute
+		{
+			public System.Type ResultType { get; set; }
+			public Delegate ExecuteOnEval { get; set; }
+			public bool IsValue { get; set; }
+		}
+
+		// ResultTransformer are usually re-usable, this is not the case of this one, which will
+		// be built for each multi-query requiring it.
+		// It also usually ends in query cache, but this is not the case either for multi-query.
+		[Serializable]
+		private class FutureResultsTransformer : IResultTransformer
+		{
+			private readonly List<BatchedQueryPostExecute> _postExecutes;
+			private int _currentIndex;
+
+			public FutureResultsTransformer(List<BatchedQueryPostExecute> postExecutes)
+			{
+				_postExecutes = postExecutes;
+			}
+
+			public object TransformTuple(object[] tuple, string[] aliases)
+			{
+				return tuple.Length == 1 ? tuple[0] : tuple;
+			}
+
+			public IList TransformList(IList collection)
+			{
+				if (_currentIndex >= _postExecutes.Count)
+					throw new InvalidOperationException(
+						$"Transformer have been called more times ({_currentIndex + 1}) than it has queries to transform.");
+
+				var postExecute = _postExecutes[_currentIndex];
+				_currentIndex++;
+				if (postExecute.ExecuteOnEval == null)
+				{
+					return collection;
+				}
+
+				var results = (IList) typeof(List<>)
+					.MakeGenericType(postExecute.ResultType)
+					.GetConstructor(System.Type.EmptyTypes)
+					.Invoke(null);
+
+				if (!postExecute.IsValue)
+				{
+					foreach (var element in (IEnumerable) postExecute.ExecuteOnEval.DynamicInvoke(collection))
+					{
+						results.Add(element);
+					}
+					return results;
+				}
+
+				// When not null on a future value, ExecuteOnEval is fetched with PostExecuteTransformer from
+				// IntermediateHqlTree through ExpressionToHqlTranslationResults, which requires a IQueryable
+				// as input and directly yields the scalar result when the query is scalar.
+				var resultElement = postExecute.ExecuteOnEval.DynamicInvoke(collection.AsQueryable());
```

Comment:
No need to type the collection for `AsQueryable` if its underlying type is already the good one, which is done in multi-query.